File: QuickRemote/android-app/app/src/main/java/com/quickremote/app/services/RemoteFrameProtocol.kt
package com.quickremote.app.services

/**
 * 远程桌面帧协议（与 PC 端一致）。
 * 帧格式：[类型 1字节][长度 4字节(小端)][载荷 N字节]
 */
object RemoteFrameProtocol {
    // ============ 帧类型 ============

    /** 视频帧（H.264 NAL unit），PC→Android。 */
    const val TYPE_VIDEO_FRAME: Byte = 0x01

    /** 鼠标事件，Android→PC。 */
    const val TYPE_INPUT_MOUSE: Byte = 0x02

    /** 键盘事件，Android→PC。 */
    const val TYPE_INPUT_KEY: Byte = 0x03

    /** 滚轮事件，Android→PC。 */
    const val TYPE_INPUT_WHEEL: Byte = 0x04

    /** 控制帧（握手/分辨率/帧率），双向。 */
    const val TYPE_CONTROL: Byte = 0x05

    /** 心跳，双向。 */
    const val TYPE_HEARTBEAT: Byte = 0x06

    /** 局域网直连认证，Android→PC：[auth_key 64B hex ASCII]。 */
    const val TYPE_AUTH: Byte = 0x07

    /** 帧头长度：1 类型 + 4 长度。 */
    const val HEADER_SIZE = 5

    /** 单帧最大载荷（16MB）。 */
    const val MAX_PAYLOAD = 16 * 1024 * 1024

    /** 编码帧头。 */
    fun makeHeader(type: Byte, length: Int): ByteArray {
        val header = ByteArray(HEADER_SIZE)
        header[0] = type
        header[1] = (length and 0xFF).toByte()
        header[2] = ((length shr 8) and 0xFF).toByte()
        header[3] = ((length shr 16) and 0xFF).toByte()
        header[4] = ((length shr 24) and 0xFF).toByte()
        return header
    }

    /** 解析帧头载荷长度（header 至少 5 字节）。 */
    fun decodeLength(header: ByteArray): Int {
        return (header[1].toInt() and 0xFF) or
            ((header[2].toInt() and 0xFF) shl 8) or
            ((header[3].toInt() and 0xFF) shl 16) or
            ((header[4].toInt() and 0xFF) shl 24)
    }
}

File: QuickRemote/pc-client/Services/RemoteFrameProtocol.cs
namespace QuickRemote.PCClient.Services;

/// <summary>
/// 远程桌面帧协议。
/// 帧格式：[类型 1字节][长度 4字节(小端)][载荷 N字节]
/// </summary>
public static class RemoteFrameProtocol
{
    // ============ 帧类型 ============

    /// <summary>视频帧（H.264 NAL unit），PC→Android。</summary>
    public const byte TYPE_VIDEO_FRAME = 0x01;

    /// <summary>鼠标事件，Android→PC。</summary>
    public const byte TYPE_INPUT_MOUSE = 0x02;

    /// <summary>键盘事件，Android→PC。</summary>
    public const byte TYPE_INPUT_KEY = 0x03;

    /// <summary>滚轮事件，Android→PC。</summary>
    public const byte TYPE_INPUT_WHEEL = 0x04;

    /// <summary>控制帧（握手/分辨率/帧率），双向。</summary>
    public const byte TYPE_CONTROL = 0x05;

    /// <summary>心跳，双向。</summary>
    public const byte TYPE_HEARTBEAT = 0x06;

    /// <summary>局域网直连认证，Android→PC：[auth_key 64B hex ASCII]。</summary>
    public const byte TYPE_AUTH = 0x07;

    /// <summary>帧头长度：1 类型 + 4 长度。</summary>
    public const int HEADER_SIZE = 5;

    /// <summary>单帧最大载荷（16MB，H.264 关键帧足够）。</summary>
    public const int MAX_PAYLOAD = 16 * 1024 * 1024;

    /// <summary>编码帧头。</summary>
    public static byte[] MakeHeader(byte type, int length)
    {
        var header = new byte[HEADER_SIZE];
        header[0] = type;
        header[1] = (byte)(length & 0xFF);
        header[2] = (byte)((length >> 8) & 0xFF);
        header[3] = (byte)((length >> 16) & 0xFF);
        header[4] = (byte)((length >> 24) & 0xFF);
        return header;
    }

    /// <summary>解析帧头长度（header 至少 5 字节）。</summary>
    public static int DecodeLength(byte[] header)
    {
        return header[1] | header[2] << 8 | header[3] << 16 | header[4] << 24;
    }
}
